### PR TITLE
C++: Add GetDelim as taint step

### PR DESCRIPTION
**Pull request by MathiasVP**
Fixes https://github.com/github/codeql-c-analysis-team/issues/58.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/GetDelim.qll [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  override predicate parameterIsAlwaysReturned(int index) { none() }
+
+  override predicate hasOnlySpecificReadSideEffects() { any() }
+
+  override predicate hasOnlySpecificWriteSideEffects() { any() }
+
+  override predicate hasSpecificWriteSideEffect(ParameterIndex i, boolean buffer, boolean mustWrite) {
+    i = [0, 1] and
+    buffer = false and
+    mustWrite = true
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/GetDelim.qll:29]:
```suggestion
    mustWrite = false
```

If I understand the manpage correctly, the first two arguments are only written to if (1) the first argument started out pointing to a `NULL` pointer or (2) a `realloc` happened.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/GetDelim.qll:29]: I'm not sure about that. On https://linux.die.net/man/3/getdelim it says:

> If *lineptr is NULL, then getline() will allocate a buffer for storing the line, which should be freed by the user program. (In this case, the value in *n is ignored.)

> Alternatively, before calling getline(), *lineptr can contain a pointer to a malloc(3)-allocated buffer *n bytes in size. If the buffer is not large enough to hold the line, getline() resizes it with realloc(3), updating *lineptr and *n as necessary.

> In either case, on a successful call, *lineptr and *n will be updated to reflect the buffer address and allocated size respectively.

The last line makes me think that the arguments are always written to.

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/models/implementations/GetDelim.qll:29]: It does say that, but what gets written? If the assignment is effectively `*p = *p` in many cases, then does that count as `mustWrite`?

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/GetDelim.qll:29]: Oh, good point! I didn't think of that. I'll accept your change.

**Pull request merged**